File: backend/schemas/pesada.py
# ============================================================
# backend/schemas/pesada.py — DTOs de la tabla central `pesadas`
# ============================================================
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict

from backend.schemas.auth import UsuarioOut
from backend.schemas.maestros import (
    VehiculoOut, ConductorOut, ProductoOut, ProveedorOut,
    DestinoOut, LoteOut, RemolqueOut, ContenedorOut,
)


class PesadaOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    numero_ticket: str
    estado: str
    tipo_pesaje: str

    fecha_entrada: Optional[datetime] = None
    fecha_captura: Optional[datetime] = None
    fecha_aprobacion: Optional[datetime] = None
    fecha_salida: Optional[datetime] = None

    peso_bruto: Optional[float] = None
    peso_tara: Optional[float] = None
    peso_neto: Optional[float] = None

    empresa_transportista: Optional[str] = None
    empresa_cliente_proveedor: Optional[str] = None
    cedula_conductor_libre: Optional[str] = None

    orden_compra: Optional[str] = None
    cantidad: Optional[float] = None
    precintos: Optional[str] = None
    observaciones: Optional[str] = None
    motivo_rechazo: Optional[str] = None
    motivo_anulacion: Optional[str] = None
    anulada: bool

    vehiculo: Optional[VehiculoOut] = None
    conductor: Optional[ConductorOut] = None
    producto: Optional[ProductoOut] = None
    proveedor: Optional[ProveedorOut] = None
    destino: Optional[DestinoOut] = None
    lote: Optional[LoteOut] = None
    remolque: Optional[RemolqueOut] = None
    contenedor: Optional[ContenedorOut] = None
    usuario_entrada: Optional[UsuarioOut] = None
    usuario_salida: Optional[UsuarioOut] = None
    aprobado_por: Optional[UsuarioOut] = None


class EntradaIn(BaseModel):
    peso_bruto: float
    vehiculo_id: int
    tipo_pesaje: str = "GENERAL"
    producto_id: Optional[int] = None
    empresa_transportista: str = ""
    empresa_cliente_proveedor: str = ""
    conductor_id: Optional[int] = None
    cedula_conductor_libre: str = ""
    proveedor_id: Optional[int] = None
    destino_id: Optional[int] = None
    lote_id: Optional[int] = None
    remolque_id: Optional[int] = None
    contenedor_id: Optional[int] = None
    observaciones: str = ""


class SalidaIn(BaseModel):
    peso_capturado: float


class RechazoIn(BaseModel):
    motivo: str


class CompletarIn(BaseModel):
    orden_compra: str = ""
    cantidad: Optional[float] = None
    precintos: str = ""
    observaciones: str = ""


class AnularIn(BaseModel):
    motivo: str


class CorteIn(BaseModel):
    observaciones: str = ""


class CorteOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    numero_corte: int
    fecha_inicio: datetime
    fecha_fin: datetime
    total_pesadas: int
    total_neto_kg: float
    observaciones: Optional[str] = None


class EstadisticasOut(BaseModel):
    en_planta: int
    completadas_hoy: int
    neto_hoy_kg: float
    total_completadas: int

File: web/src/types/pesada.ts
// Subconjunto de backend/schemas/pesada.py → PesadaOut: solo los campos
// que esta web usa. El backend manda bastantes más (pesos de guía,
// precintos, observaciones, etc.); agregarlos acá a medida que hagan
// falta, no "por las dudas".

export interface VehiculoDePesada {
  placa: string
  descripcion: string | null
  tipo?: string | null
}

export interface ProductoDePesada {
  codigo: string
  nombre: string
}

/** backend/schemas/maestros.py → ConductorOut (lo que usa el detalle). */
export interface ConductorDePesada {
  nombre: string
  documento: string
  tipo_documento?: string | null
}

/** ProveedorOut y EmpresaTransportistaOut tienen la misma forma. */
export interface EmpresaDePesada {
  codigo: string
  nombre: string
  rif?: string | null
}

export interface DestinoDePesada {
  codigo: string
  nombre: string
}

/** backend/schemas/auth.py → UsuarioOut, reducido a lo que se muestra. */
export interface UsuarioDePesada {
  id: number
  username: string
  nombre_completo: string
}

/** Los cuatro estados que muestra el tablero (hay más: rechazado, anulado). */
export type EstadoTablero = 'en_planta' | 'pendiente_aprobacion' | 'aprobado' | 'completado'

export interface Pesada {
  id: number
  numero_ticket: string
  estado: string
  tipo_pesaje: string

  fecha_entrada: string | null
  fecha_captura: string | null
  fecha_aprobacion: string | null
  fecha_salida: string | null

  peso_entrada: number | null
  peso_bruto: number | null
  peso_tara: number | null
  peso_neto: number | null

  codigo_viaje: string | null
  peso_guia: number | null
  bultos: number | null

  auto_aprobado: boolean
  es_manual: boolean

  empresa_transportista: string | null
  empresa_cliente_proveedor: string | null

  vehiculo: VehiculoDePesada | null
  producto: ProductoDePesada | null

  // --- Detalle completo (GET /pesadas/{id}, también el kardex) ---
  // Opcionales a propósito: los listados del tablero y los datos de
  // prueba no siempre los traen, y el detalle tiene que tolerar su
  // ausencia igual que un null (pesadas viejas, anteriores al campo).
  peso_final?: number | null
  anulada?: boolean
  cedula_conductor_libre?: string | null
  procedencia?: string | null
  orden_compra?: string | null
  cantidad?: number | null
  precintos?: string | null
  observaciones?: string | null
  motivo_rechazo?: string | null
  /** Por qué aprobó Costos (opcional). */
  comentario_aprobacion?: string | null
  motivo_anulacion?: string | null
  fecha_anulacion?: string | null

  conductor?: ConductorDePesada | null
  proveedor?: EmpresaDePesada | null
  transportista?: EmpresaDePesada | null
  destino?: DestinoDePesada | null

  // Quién hizo cada etapa: con esto y las fechas se arma el recorrido
  // (el backend no guarda un historial de auditoría aparte).
  usuario_entrada?: UsuarioDePesada | null
  usuario_salida?: UsuarioDePesada | null
  aprobado_por?: UsuarioDePesada | null
  usuario_completado?: UsuarioDePesada | null
  anulado_por?: UsuarioDePesada | null
}
